tokio-io/src/io/buf_stream.rs

```diff
@@ -0,0 +1,18 @@
+use crate::io::{BufReader, BufWriter};
+use crate::{AsyncRead, AsyncWrite};
+
+/// Wraps a type that is [`AsyncWrite`] and [`AsyncRead`], and buffers its input and output.
+///
+/// It can be excessively inefficient to work directly with something that implements [`AsyncWrite`]
+/// and [`AsyncRead`]. For example, every `write`, however small, has to traverse the syscall
+/// interface, and similarly every read has to do the same. The [`BufWriter`] and [`BufReader`]
+/// types aid with these problems respectively, but do so in only one direction. `BufStream` wraps
+/// one in the other so that both directions are buffered. See their documentation for details.
+pub type BufStream<RW> = BufReader<BufWriter<RW>>;
+
+/// Wrap a type in both [`BufWriter`] and [`BufReader`].
+///
+/// See the documentation for those types and [`BufStream`] for details.
+pub fn buffer<RW: AsyncRead + AsyncWrite>(stream: RW) -> BufStream<RW> {
```

Comment:
Should there also be an extension trait method for this somewhere?

Reply:
We could probably add one to `AsyncReadExt` and `AsyncWriteExt` 👍

Reply:
I decided not to do this because a) there aren't extension methods for making `BufReader` and `BufWriter`, and b) the name `buffer` on either extension trait should probably be reserved for one-way buffering. If you want I could add both of those, and then we'd have to find a good name for a method that, given an `AsyncReadExt where Self: AsyncWrite` or `AsyncWriteExt where Self: AsyncRead`, gives you `BufStream`.

Reply:
`buffered_duplex`, `buffered_read` and `buffered_write` etc?

Reply:
Hmm, it's actually also a little awkward, because those methods would (presumably) consume `self` rather than take `&mut self`, which means they also require `Self: Sized`. Are we okay with that?

Reply:
I'd say the only weird thing is that we have a free fn here and it seems we have been going with the non free fn pattern, beyond that I don't mind. I think that is fine to take `Sized`.

Reply:
Yeah, the free fn is just because `BufStream` isn't a newtype, but a type alias, so I can't add a `new` method to it. I _could_ change it to a type, though then I'd also have to forward all the traits manually. If you'd prefer that though, I could do it. I've added the `Ext` trait methods now (I called them `buffered` and `buffered_duplex`).

Reply:
Tbh, that may be the right move actually now that I think about it, type alias as its a nice trick feels a bit awkward. So newtype is prob better.

Reply:
Done in latest commit.